File: Sources/Services/RemoteControl/RemoteAuditLog.swift
// MARK: - RemoteAuditLog
// Structured audit logging for all Remote Control security-relevant events.
// macOS 14+, Swift 5.10

import Foundation
import OSLog

/// Centralised audit logging for the Remote Control server.
///
/// Every security-relevant event (authentication, connection, terminal input)
/// is logged through this enum so that audit trails are consistent and
/// discoverable in Console.app under the `RemoteControl` category.
///
/// Log levels:
/// - `.info` -- successful operations (connect, disconnect, input).
/// - `.error` -- failed authentication, lockout events.
/// - `.fault` -- server disabled due to security policy violation.
enum RemoteAuditLog {

    // MARK: - Authentication

    /// Log an authentication attempt (PIN validation).
    ///
    /// - Parameters:
    ///   - ip: The client's IP address.
    ///   - success: Whether the PIN was accepted.
    static func authAttempt(ip: String, success: Bool) {
        if success {
            Logger.remoteControl.info("[AUDIT] Auth success from IP: \(ip, privacy: .public)")
        } else {
            Logger.remoteControl.error("[AUDIT] Auth failure from IP: \(ip, privacy: .public)")
        }
    }

    /// Log a rate-limit lockout event.
    ///
    /// - Parameters:
    ///   - ip: The locked-out IP address.
    ///   - global: Whether this triggered a global (all-IP) lockout.
    static func authLockout(ip: String, global: Bool) {
        if global {
            Logger.remoteControl.fault("[AUDIT] GLOBAL LOCKOUT triggered by IP: \(ip, privacy: .public)")
        } else {
            Logger.remoteControl.error("[AUDIT] IP lockout: \(ip, privacy: .public)")
        }
    }

    // MARK: - Device Lifecycle

    /// Log a successful device connection.
    ///
    /// - Parameters:
    ///   - device: The authenticated remote device.
    ///   - sessionId: The terminal session the device attached to.
    static func deviceConnect(device: RemoteDevice, sessionId: UUID) {
        Logger.remoteControl.info(
            "[AUDIT] Device connected: \(device.displayName, privacy: .public) (\(device.ipAddress, privacy: .public)) -> session \(sessionId)"
        )
    }

    /// Log a device disconnection.
    ///
    /// - Parameters:
    ///   - deviceId: The disconnected device's identifier.
    ///   - reason: Human-readable reason (e.g. "idle_timeout", "kicked_by_host").
    static func deviceDisconnect(deviceId: UUID, reason: String) {
        Logger.remoteControl.info(
            "[AUDIT] Device disconnected: \(deviceId) reason=\(reason, privacy: .public)"
        )
    }

    // MARK: - Terminal I/O

    /// Log terminal input relayed from a remote device.
    ///
    /// Only the byte length is logged -- **never** the content itself, which
    /// could contain passwords, API keys, or other secrets.
    ///
    /// - Parameters:
    ///   - deviceId: The device that sent the input.
    ///   - sessionId: The terminal session that received the input.
    ///   - length: Number of UTF-8 bytes in the input payload.
    static func terminalInput(deviceId: UUID, sessionId: UUID, length: Int) {
        Logger.remoteControl.info(
            "[AUDIT] Input: device=\(deviceId) session=\(sessionId) bytes=\(length)"
        )
    }

    // MARK: - Server Lifecycle

    /// Log the server being disabled due to a security policy violation.
    ///
    /// - Parameter reason: Description of the security event.
    static func serverDisabled(reason: String) {
        Logger.remoteControl.fault(
            "[AUDIT] Server DISABLED: \(reason, privacy: .public)"
        )
    }
}

File: shared/src/commonMain/kotlin/studio/vibe/shared/service/remote/RemoteAuditLog.kt
package studio.vibe.shared.service.remote

import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid
import studio.vibe.shared.model.RemoteDevice

// ── PlatformLogger Interface ──────────────────────────────────────────────────

/**
 * Platform-injectable structured logger.
 *
 * Platform implementations map these calls to the appropriate system logging facility:
 * - macOS/iOS: `os_log` / `Logger`
 * - JVM: `java.util.logging` or SLF4J
 * - Browser: `console.*`
 */
interface PlatformLogger {
    fun info(tag: String, message: String)
    fun warning(tag: String, message: String)
    fun error(tag: String, message: String)
}

// ── RemoteAuditLog ────────────────────────────────────────────────────────────

/**
 * Centralised audit logging for all Remote Control security-relevant events.
 *
 * Every security event (authentication attempt, device connection/disconnection,
 * terminal input relay, server shutdown) is channelled through this class so that
 * audit trails are consistent and filterable by the `REMOTE_CONTROL` tag.
 *
 * Log levels:
 * - [PlatformLogger.info]    — successful operations (connect, disconnect, input relay).
 * - [PlatformLogger.warning] — per-IP lockout.
 * - [PlatformLogger.error]   — failed authentication attempts, global lockout.
 *
 * Terminal content is **never** logged — only the byte length of input payloads.
 * This prevents passwords, API keys, and other secrets from appearing in logs.
 */
@OptIn(ExperimentalUuidApi::class)
class RemoteAuditLog(private val logger: PlatformLogger) {

    companion object {
        private const val TAG = "REMOTE_CONTROL"
    }

    // ── Authentication ────────────────────────────────────────────────────────

    /**
     * Log a PIN authentication attempt.
     *
     * @param ip      The client's IP address.
     * @param success Whether the PIN was accepted.
     */
    fun authAttempt(ip: String, success: Boolean) {
        if (success) {
            logger.info(TAG, "[AUDIT] Auth success from IP: $ip")
        } else {
            logger.error(TAG, "[AUDIT] Auth failure from IP: $ip")
        }
    }

    /**
     * Log a rate-limit lockout event.
     *
     * @param ip     The locked-out IP address.
     * @param global Whether this triggered a global (all-IP) lockout.
     */
    fun authLockout(ip: String, global: Boolean) {
        if (global) {
            logger.error(TAG, "[AUDIT] GLOBAL LOCKOUT triggered by IP: $ip")
        } else {
            logger.warning(TAG, "[AUDIT] IP lockout: $ip")
        }
    }

    // ── Device Lifecycle ──────────────────────────────────────────────────────

    /**
     * Log a successful device connection.
     *
     * @param device    The authenticated remote device.
     * @param sessionId The terminal session the device attached to.
     */
    fun deviceConnect(device: RemoteDevice, sessionId: Uuid) {
        logger.info(
            TAG,
            "[AUDIT] Device connected: ${device.displayName} (${device.ipAddress}) -> session $sessionId",
        )
    }

    /**
     * Log a device disconnection.
     *
     * @param deviceId The disconnected device's identifier.
     * @param reason   Human-readable reason (e.g. `"idle_timeout"`, `"kicked_by_host"`).
     */
    fun deviceDisconnect(deviceId: Uuid, reason: String) {
        logger.info(TAG, "[AUDIT] Device disconnected: $deviceId reason=$reason")
    }

    // ── Terminal I/O ──────────────────────────────────────────────────────────

    /**
     * Log terminal input relayed from a remote device.
     *
     * Only the **byte count** is logged — the content is never logged.
     * This prevents passwords, API keys, or other secrets from appearing in audit logs.
     *
     * @param deviceId  The device that sent the input.
     * @param sessionId The terminal session that received the input.
     * @param length    Number of bytes in the input payload.
     */
    fun terminalInput(deviceId: Uuid, sessionId: Uuid, length: Int) {
        logger.info(TAG, "[AUDIT] Input: device=$deviceId session=$sessionId bytes=$length")
    }

    // ── Server Lifecycle ──────────────────────────────────────────────────────

    /**
     * Log the server being disabled due to a security policy violation.
     *
     * @param reason Description of the triggering security event.
     */
    fun serverDisabled(reason: String) {
        logger.error(TAG, "[AUDIT] Server DISABLED: $reason")
    }
}
